src/shogun/multiclass/KNNSolver.h

```diff
@@ -0,0 +1,195 @@
+ * include most of Variables which been used all the other Solver and methods been used to choose 
+ * the index of the most frequent class.
+ */
+
+class CDistanceMachine;
+class CKNNSolver : public CDistanceMachine
+{
+	public:
+		/** default constructor */
+		CKNNSolver(): CDistanceMachine() { init(); }
+
+		virtual ~CKNNSolver() { /* nothing to do */ }
+
+		CKNNSolver(int32_t k, float64_t q, int32_t num_classes, int32_t min_label, SGVector<int32_t> train_labels);
+
+		/** compute the histogram of class outputs of the k nearest 
+		 * neighbors to a test vector and return the index of the most frequent class
+		 *
+		 * @param classes vector used to store the histogram
+		 * @param train_lab class indices of the training data. If the cover
+		 * tree is not used, the elements are ordered by increasing distance
+		 * and there are elements for each of the training vectors. If the cover
+		 * tree is used, it contains just m_k elements not necessary ordered.+		 *
+		 * @return index of the most frequent class, class detected by KNN
+		 */
+		int32_t choose_class(float64_t* classes, int32_t* train_lab);
+
+		/** compute the histogram of class outputs of the k nearest neighbors +		 *  to a test vector, using k from 1 to m_k, and write the most frequent
+		 *  class for each value of k in output, using a distance equal to step
+		 *  between elements in the output array
+		 *
+		 * @param output return value where the most frequent classes are written
+		 * @param classes vector used to store the histogram
+		 * @param train_lab class indices of the training data; no matter the cover tree
+		 * is used or not, the neighbors are ordered by distance to the test vector
+		 * in ascending order
+		 * @param step distance between elements to be written in output
+		 */
+		void choose_class_for_multiple_k(int32_t* output, int32_t* classes, int32_t* train_lab, int32_t step);
+
+		/**
+		 * classify objects based on different way to implementation.
+		 */
+		virtual CMulticlassLabels* classify_objects(CDistance* d, int32_t num_lab, int32_t* train_lab, float64_t* classes)
+		{ 
+			CMulticlassLabels* output=NULL;
+			return output;
+		}
+
+		/**
+		 * classify all objects based on different way to implementation.
+		 */
+		virtual int32_t* classify_objects_k(CDistance* d, int32_t num_lab, int32_t* train_lab, int32_t* classes)
+		{ 
+			int32_t* output = NULL;
+			return output;
+		}
+
+		/** @return object name */
+		virtual const char* get_name() const { return "KNNSolver"; }
+
+		void init();
+
+	protected:
+		/// the k parameter in KNN
+		int32_t m_k;
+
+		/// parameter q of rank weighting
+		float64_t m_q;
+
+		/// number of classes (i.e. number of values labels can take)
+		int32_t m_num_classes;
+
+		/// smallest label, i.e. -1
+		int32_t m_min_label;
+ 
+		/** the actual trainlabels */
+		SGVector<int32_t> m_train_labels;
+};
+
+class CBRUTESolver : public CKNNSolver
```

Comment:
man, not only the implementation should be in a separate file but as well the definition.

moreover, the naming of these different KNN solvers are very misleading, as `CBRUTESolver` is not a general brute force solver, it's just solves the KNN problem itself in a bruteforce manner... please incorporate this in the naming of the class as well the placement of the files.
same is true for the rest of the solvers...